peps/pep-0800.rst

```diff
@@ -0,0 +1,508 @@
+
+The concept of "solid bases" has been part of the CPython implementation for a long time;
+the concept dates back to `a 2001 commit <https://github.com/python/cpython/commit/6d6c1a35e08b95a83dbe47dbd9e6474daff00354>`__.
+Nevertheless, the concept has received little attention in the documentation.
+Although details of the mechanism are closely tied to CPython's internal object representation,
+it is useful to explain at a high level how and why CPython works this way.
+
+Every object in CPython is essentially a pointer to a C struct, a contiguous piece of memory that
+contains information about the object. Some information is managed by the interpreter and shared
+by many or all objects, such as a reference to the type of the object, and the attribute ``__dict__``
+for user-defined objects. Some classes contain additional information that is specific to that class.
+For example, user-defined classes with ``__slots__`` contain a place in memory for each slot,
+and the built-in ``float`` class contains a C ``double`` value that stores the value of the float.
+Code that interacts with these classes usually assumes a certain memory layout: C code that
+interacts with a ``float`` expects to find the value at a particular offset in the object's memory.
+
+When a child class is created, CPython must create a memory layout for the new class that
+is compatible with all of its parent classes. For example, when a child class of ``float``
+is created, it must be possible to pass instances of the child class to C code that interacts
+directly with the underlying struct for the ``float`` class. Therefore, such a subclass must store
+the ``double`` value at the same offset as the parent ``float`` class does. It may, however, add
+additional fields at the end of the struct. CPython knows how to do this with the ``__dict__``
+attribute, which is why it is possible to create a child class of ``float`` that adds a ``__dict__``.
+
+However, there is no way to combine a ``float``, which must have a ``double`` in its struct,
+with another C type like ``int``, which stores different data at the same spot. Therefore,
+a common subclass of ``float`` and ``int`` cannot exist. We say that ``float`` and ``int``
+are solid bases.
+
+Classes implemented in C are solid bases if if they have an underlying struct that stores
+data at a fixed offset, and that struct is different from the struct of its parent class.
+C classes may also store a variable-size array of data (such as the contents of a string);
+if this differs from the parent class, the class also becomes a solid base.
+CPython's implementation deduces this from the :c:member:`~PyTypeObject.tp_itemsize`
+and :c:member:`~PyTypeObject.tp_basicsize` fields of the type object, which are also
+accessible from Python code as the undocumented attributes ``__itemsize__`` and ``__basicsize__``
+on type objects.
+
+Similarly, classes implemented in Python are solid bases if they have ``__slots__``, because
```

Comment:
Worth linking to the datamodel docs for slots here?

Reply:
```suggestion
Similarly, classes implemented in Python are solid bases if they have :py:attr:`~object.__slots__`, because
```